### PR TITLE
[SofaKernel] ADD: add polynomial springs force fields

### PR DESCRIPTION
**Pull request by sergeiNikolaev**
This components allow to simulate springs with Polynomial stress strain behavior.
The general force computation looks like on image below:
GenImage

Despite the fact I am using only polynomial springs, in general it is possible to set any first order smooth function as stress-strain relation.
The polynomial parameters are set as two arrays. d_polynomialDegree describe set of polynomial degrees for every spring. d_polynomialStiffness describe set of polynomial coefficients sequentially combined in one vector. The coefficients are put from smaller degree to bigger one, and the free coefficient is always zero (since for no strain we have no stress). For examples the coefficients for polynomials [3,2,4] will be put as [a1,a2,a3,b1,b2,c1,c2,c3,c4].
The dedication of jacobian matrix for PolynomialSpringForceField is given below
Pol1
Pol2
Pol3
Pol4
I guess the dedication has to be added as documentation, but I do not know how to add it.

For RestShapeSpring I added an exponetial addition to the denominator to avoid Nan problems for cases when spring has a zero length. As a result, the stress simulation is shifted compared with polynomial values, but it keeps its nonlinearity. Here is the dedication of a derivative
RestPol1
RestPol2
RestPol3
RestPol4

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by sergeiNikolaev** [comment]
> Thanks @sergeiNikolaev for this contribution and making this nice effort of bringing back this work.
> 
> * Is any test available with these new classes?
> * is there any example scene?

Thank you for your answer. Yes, there is an example scene, which I also use as a test. But to add it to Sofa I have to remove plugin dependencies from it.

**Comment by fredroy** [comment]
Amazing work @sergeiNikolaev !
However, as it seems you want into SofaCommon, we have to be a bit strict on the guidelines 👍 
So could you: 
 - remove all useless commented code?
 - set prefix for Data<> (d_) and member (m_) ?
 - remove serr with msg_error
 - maybe uncomment all the msg_info (as it is not print if you dont enable printLog)
 - would be cool to use pragma once instead of the preprocessor guards (ifdef)
Thanks !

**Comment by adagolodjo** [comment]
Hi @sergeiNikolaev, thanks for this contribution.  I just stared reading your equations and I will probably have somes question in order to understand more about it.  
1) First the reference (paper where you got this force equation from) of the equation 1
2) In the first image why you have absolut value on the current lenght (∆l) if this is a length it probably positif rigth?
3) Why current length ∆l = (∆x^2 + ∆y^2 +∆z^2 )^(1/2) it looks more like the formula of the displacement right ?

**Comment by sergeiNikolaev** [comment]
Hello.
Thank you for your comments, guys.

@Younesssss , answering to your questions:

> 1. First the reference (paper where you got this force equation from) of the equation 1

It is a general spring equation. You could find it for example here: Nealen A. et al "Physically Based Deformable Models in Computer Graphics" 
https://matthias-research.github.io/pages/publications/egstar2005.pdf or in StiffSpringForceField SOFA component.
A spring force contains two parts: value and direction. Compared with linear spring the value now is computed like a result of polynomial. The computation of direction remains the same. The S (cross-section area) technically has to describe the characteristics of material, but since it depends on sample shape we are trying to simulate, finally we could say that we compute force intensity instead of object force and, thus, have to recompute it based on sample shape and amount of springs we are using for simulation.

> 2. In the first image why you have absolute value on the current length (∆l) if this is a length it probably positif rigth?

Yes, the length is always positive. I added absolute value since square root, in general, might also mean negative value. 

> 3. Why current length ∆l = (∆x^2 + ∆y^2 +∆z^2 )^(1/2) it looks more like the formula of the displacement right ?

Yes, it is a displacement, but this displacement describes a distance between spring endpoints (actually, p1 - p2), not the displacement between the object and its rest shape. Therefore, I marked these displacements with notation ∆. Spring length is a norm of this distance.

**Comment by sergeiNikolaev** [comment]
> Amazing work @sergeiNikolaev !
> However, as it seems you want into SofaCommon, we have to be a bit strict on the guidelines 
> So could you:
> 
> * remove all useless commented code?
> * set prefix for Data<> (d_) and member (m_) ?
> * remove serr with msg_error
> * maybe uncomment all the msg_info (as it is not print if you dont enable printLog)
> * would be cool to use pragma once instead of the preprocessor guards (ifdef)
>   Thanks !

@fredroy 
I have submitted the correction to the code.

**Comment by sergeiNikolaev** [comment]
> Thanks @sergeiNikolaev for this contribution and making this nice effort of bringing back this work.
> 
> * Is any test available with these new classes?
> * is there any example scene?

@hugtalbot
Here is the test to verify polynomial springs (Python2 scripts).

[polynomial_spring_test.zip](https://github.com/sofa-framework/sofa/files/4540453/polynomial_spring_test.zip)

The test "spring_verification.py" depicts two points (p1, p2) connected with a spring. One point (p1) is fixed, another (p2) is moving in space. The external impact (f_external) to the system could be applied using LinearForceField (you need to make target_force as a Data<> object in order to collect its information, but I have this modification) or ControlPoint (third point (p_control), attached with a linear rest shape spring). In case of a second option, the external force is computed as (p_control - p2)*k_rest_shape_spring.
During the simulation process, the position "p2" and forces "f2" for point p2, and also the external force or the position of the control point are saved to data files.
The script "spring_verification_draw_result.py" loads all data. The strain for spring is computed as 
(norm(p2-p1) - l0) / l0 for PolynomialSpring or as norm(p2-p1) / l0 for RestShapePolynomialSpring.
I use static solver with many iterations to make spring force equal to external force, but I guess it is also possible to compute the spring force as difference  f2 - f_external.
Based on this data, a force-strain relation is plotted and compared with polynomial values.
And this is the end of the test.

NOTE: Please, keep in mind, that configuration parameters have to be the same for both scripts, in order to get a valid result.

**Comment by sergeiNikolaev** [comment]
Ah, yes. The comparison of a force-strain relation is done on the last figure (figure 5). All other figures just show the tracking of different parameters through the process.

**Comment by damienmarchal** [comment]
@guparan, @Younesssss, @hugtalbot  any volunteer to integrate the test in the normal test framework so we can merge this asap ?

**Comment by sergeiNikolaev** [comment]
Hello.

> Hi @sergeiNikolaev
> thank you for sharing your zip test. I think we will be able to work it out from here.
> TODO for me:
> 
> * Need a test in c++ for checking new functionality, to be implemented from py example.
> * Add an example in SOFA.

My question is what test it is possible to write for Polynomial springs?
Personally me, I verified them by generating some data using the scene from .zip and then compare it visually with the polynomial result. But, I am not sure it is possible to perform the same stuff as a functional test.

**Comment by sergeiNikolaev** [comment]
> Hi @sergeiNikolaev
> Sorry for the latency on this topic, tests can be made so that we create a simple scene in c++ imposing rest shape and an imposed displacement. Thus, the resulting force should be analytically defined.
> By the way do you have any latex file for the polynomial equations?

Thank you for your answer @hugtalbot.
Regarding your question, I do not, sorry.
I wrote these equations since for me it is easier to process the stuff like this with a "writing memory". And then I decided that it will be enough just to scan them and send to you.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 1]
```diff
@@ -0,0 +1,598 @@
...
+    {
+        recomputeIndices();
+    }
+
+    Real compressionValue = d_compressible.getValue() ? -1.0 : 0.0;
+
+    msg_info() << "\n\nNew step:\n";
+    if (d_polynomialDegree.getValue().size() != m_firstObjectIndices.size())
+    {
+        msg_warning() << "WARNING : stiffness is not defined on each point, first stiffness is used";
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:215]: Redit warning.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 2]
```diff
@@ -0,0 +1,598 @@
...
+PolynomialSpringsForceField<DataTypes>::PolynomialSpringsForceField()
+    : PolynomialSpringsForceField(nullptr, nullptr)
+{
+}
+
+
+template<class DataTypes>
+PolynomialSpringsForceField<DataTypes>::PolynomialSpringsForceField(MechanicalState* mstate1, MechanicalState* mstate2)
+    : Inherit(mstate1, mstate2)
+    , d_firstObjectPoints(initData(&d_firstObjectPoints, "firstObjectPoints", "points related to the first object"))
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:62]: Why not use the same naming as the ones used in SpringsForceField ? Renaming things is the best way to annoy users.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 3]
```diff
@@ -0,0 +1,598 @@
...
+    {
+        msg_info() << "default case for object 2: points = all points";
+        for (unsigned int i = 0; i < (unsigned)this->mstate2->getSize(); i++) {
+            m_secondObjectIndices.push_back(i);
+        }
+    }
+
+    if (m_firstObjectIndices.size() > m_secondObjectIndices.size())
+    {
+        msg_error() << "Error : the dimention of the source and the targeted points are different ";
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:180]: msg_error() + Error.

File: SofaKernel/modules/SofaDeformable/PolynomialRestShapeSpringsForceField.h [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+    Data<bool> d_drawSpring;                      ///< draw Spring
+    Data<defaulttype::RGBAColor> d_springColor;
+    Data<float> d_showIndicesScale;
+
+    Data<VecReal> d_zeroLength;       /// Springs initial lengths
+    Data<double> d_smoothShift;
+    Data<double> d_smoothScale;
+
+    SingleLink<PolynomialRestShapeSpringsForceField<DataTypes>, sofa::core::behavior::MechanicalState<DataTypes>,
+        BaseLink::FLAG_STOREPATH|BaseLink::FLAG_STRONGLINK> d_restMState;
```
Comment (hugtalbot) [inline SofaKernel/modules/SofaDeformable/PolynomialRestShapeSpringsForceField.h:106]: indent

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 5]
```diff
@@ -0,0 +1,598 @@
...
+                msg_info() << "Spring zero length: " << m_initialSpringLength[i];
+                m_computeSpringsZeroLength[i] = 0;
+            }
+            m_weightedCoordinateDifference[i] = m_weightedCoordinateDifference[i] / m_springLength[i];
+            msg_info() << "Weighted coordinate difference: " << m_weightedCoordinateDifference[i];
+
+            m_strainValue[i] = std::fabs(m_springLength[i] - m_initialSpringLength[i]) / m_initialSpringLength[i];
+            double forceValue = PolynomialValue(0, m_strainValue[i]);
+            m_strainSign[i] = m_springLength[i] - m_initialSpringLength[i] >= 0 ? 1.0 : compressionValue;
+            msg_info() << "Strain sign: " << m_strainSign[i];
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:239]: Make a single message msg_info() instead of several ones so this correspond to only one message in the GUI.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 6]
```diff
@@ -0,0 +1,598 @@
...
+                m_initialSpringLength[i] = m_springLength[i];
+                msg_info() << "Spring zero length: " << m_initialSpringLength[i];
+                m_computeSpringsZeroLength[i] = 0;
+            }
+            m_weightedCoordinateDifference[i] = m_weightedCoordinateDifference[i] / m_springLength[i];
+
+            m_strainValue[i] = std::fabs(m_springLength[i] - m_initialSpringLength[i]) / m_initialSpringLength[i];
+            double forceValue = PolynomialValue(i, m_strainValue[i]);
+            m_strainSign[i] = m_springLength[i] - m_initialSpringLength[i] >= 0 ? 1.0 : compressionValue;
+            msg_info() << "Strain sign: " << m_strainSign[i];
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:272]: Same here..multiple msg_info

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 7]
```diff
@@ -0,0 +1,598 @@
...
+        }
+    }
+}
+
+
+
+template<class DataTypes>
+void PolynomialSpringsForceField<DataTypes>::ComputeJacobian(unsigned int stiffnessIndex, unsigned int springIndex)
+{
+    msg_info() << "\n\nCompute derivative: ";
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:290]: same here.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 8]
```diff
@@ -0,0 +1,598 @@
...
+    {
+        for(unsigned int secondIndex = 0; secondIndex < m_dimension; secondIndex++)
+        {
+            jacobMatrix[firstIndex][secondIndex] = (polynomialDerivativeRes - polynomialForceRes) *
+                    m_weightedCoordinateDifference[springIndex][firstIndex] * m_weightedCoordinateDifference[springIndex][secondIndex];
+        }
+        jacobMatrix[firstIndex][firstIndex] += polynomialForceRes;
+    }
+
+    for(unsigned int firstIndex = 0; firstIndex < m_dimension; firstIndex++)
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:316]: The msg_info is done in the for loop... so if there is no printing we are still iterating.
Add a condition like this one:  if(notMuted()){ the loops... }

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 9]
```diff
@@ -0,0 +1,598 @@
...
+template<class DataTypes>
+void PolynomialSpringsForceField<DataTypes>::addDForce(const core::MechanicalParams* mparams, DataVecDeriv& data_df1, DataVecDeriv& data_df2,
+                                                       const DataVecDeriv& data_dx1, const DataVecDeriv& data_dx2)
+{
+    VecDeriv&        df1 = *data_df1.beginEdit();
+    VecDeriv&        df2 = *data_df2.beginEdit();
+    const VecDeriv&  dx1 =  data_dx1.getValue();
+    const VecDeriv&  dx2 =  data_dx2.getValue();
+
+    msg_info() << "[" <<  this->getName() << "]: addDforce";
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:335]: why manually doing this->getName() ?

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 10]
```diff
@@ -0,0 +1,598 @@
...
+    vparams->drawTool()->saveLastState();
+
+    const VecCoord& p1 =this->mstate1->read(core::ConstVecCoordId::position())->getValue();
+    const VecCoord& p2 =this->mstate2->read(core::ConstVecCoordId::position())->getValue();
+
+    const VecIndex& firstObjectIndices = d_firstObjectPoints.getValue();
+    const VecIndex& secondObjectIndices = d_secondObjectPoints.getValue();
+
+    std::vector< defaulttype::Vector3 > points;
+    for (unsigned int i = 0; i < firstObjectIndices.size(); i++)
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:365]: use c++11 for each loop instead of indices.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 11]
```diff
@@ -0,0 +1,598 @@
...
+
+    if (d_showArrowSize.getValue() == 0 || d_drawMode.getValue() == 0)
+    {
+        vparams->drawTool()->drawLines(points, 1, d_springColor.getValue());
+    }
+    else if (d_drawMode.getValue() == 1)
+    {
+        const unsigned int numLines = points.size() / 2;
+        for (unsigned int i = 0; i < numLines; ++i) {
+            vparams->drawTool()->drawCylinder(points[2*i+1], points[2*i], d_showArrowSize.getValue(), d_springColor.getValue());
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:381]: Put the drawTool() out of the loop. 
as well as all the getValue() (because this trigs the data graph update each time)

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+        }
+    }
+    else if (d_drawMode.getValue() == 2)
+    {
+        const unsigned int numLines = points.size() / 2;
+        for (unsigned int i = 0; i < numLines; ++i) {
+            vparams->drawTool()->drawArrow(points[2*i+1], points[2*i], d_showArrowSize.getValue(), d_springColor.getValue());
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:388]: Put the drawTool() out of the loop. 
as well as all the getValue() (because this trigs the data graph update each time)

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 13]
```diff
@@ -0,0 +1,598 @@
...
+        }
+    }
+
+    // draw connected point indices
+    defaulttype::Vec4f color(1.0, 1.0, 1.0, 1.0);
+
+    Real scale = (vparams->sceneBBox().maxBBox() - vparams->sceneBBox().minBBox()).norm() * d_showIndicesScale.getValue();
+
+    helper::vector<defaulttype::Vector3> positions;
+    for (size_t i = 0; i < firstObjectIndices.size(); i++) {
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:398]: Use c++1 for each instead of using indices.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 14]
```diff
@@ -0,0 +1,598 @@
...
+        const unsigned int index = firstObjectIndices[i];
+        positions.push_back(defaulttype::Vector3(p1[index][0], p1[index][1], p1[index][2] ));
+    }
+    for (size_t i = 0; i < secondObjectIndices.size(); i++) {
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:402]: Use c++1 for each instead of using indices.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 15]
```diff
@@ -0,0 +1,598 @@
...
+    vparams->drawTool()->restoreLastState();
+}
+
+
+
+template<class DataTypes>
+void PolynomialSpringsForceField<DataTypes>::addKToMatrix(const core::MechanicalParams* mparams,
+                                                          const sofa::core::behavior::MultiMatrixAccessor* matrix )
+{
+    msg_info() << "[" <<  this->getName() << "]: addKToMatrix";
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:418]: why doing manually the [this->getName()] ? 
printing addKToMatrix seems to be a debugging feature... and is not supposed to be a real message for the user. do use dmsg_info() .

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 16]
```diff
@@ -0,0 +1,598 @@
...
+
+    sofa::helper::AdvancedTimer::stepBegin("restShapeSpringAddKToMatrix");
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:420]: Use the RAII version of the timer instead.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 17]
```diff
@@ -0,0 +1,598 @@
...
+    unsigned int secondIndex = 0;
+
+    if (this->mstate1 == this->mstate2)
+    {
+        sofa::core::behavior::MultiMatrixAccessor::MatrixRef mref = matrix->getMatrix(this->mstate1);
+        if (!mref) return;
+        sofa::defaulttype::BaseMatrix* mat = mref.matrix;
+        unsigned int offset = mref.offset;
+
+        for (unsigned int index = 0; index < m_firstObjectIndices.size(); index++)
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:498]: use the c++11 for each loop.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 18]
```diff
@@ -0,0 +1,598 @@
...
+        }
+    }
+}
+
+
+
+template<class DataTypes>
+double PolynomialSpringsForceField<DataTypes>::PolynomialValue(unsigned int springIndex, double strainValue)
+{
+    helper::ReadAccessor<Data<VecReal>> vPolynomialStiffness = d_polynomialStiffness;
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:557]: use the getReadAccessor instead of implicit conversion because it does not requires the coder to know the left side type. 
Eg: 
```cpp
auto polynomialStiffness = getReadAccessor(d_polynomialStiffness);
```

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 19]
```diff
@@ -0,0 +1,598 @@
...
+    helper::ReadAccessor<Data<helper::vector<unsigned int> >> vPolynomialDegree = d_polynomialDegree;
+
+    msg_info() << "Polynomial data: ";
+    double highOrderStrain = 1.0;
+    double result = 0.0;
+    for (size_t degreeIndex = 0; degreeIndex < vPolynomialDegree[springIndex]; degreeIndex++) {
+        highOrderStrain *= strainValue;
+        result += vPolynomialStiffness[m_polynomialsMap[springIndex][degreeIndex]] * highOrderStrain;
+        msg_info() << "Degree:" << (degreeIndex + 1) << ", result: " << result;
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:566]: Don't do msg_info() in a for loop.... it is better to have a single msg_info after the loop...and concatenating all the "degree" into a single stringstream. Both faster and clearer.

If the aim of this loop is to print something then put everything in a if(notMuted()){}

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl [hunk 20]
```diff
@@ -0,0 +1,598 @@
...
+template<class DataTypes>
+double PolynomialSpringsForceField<DataTypes>::PolynomialDerivativeValue(unsigned int springIndex, double strainValue)
+{
+    helper::ReadAccessor<Data<VecReal>> vPolynomialStiffness = d_polynomialStiffness;
+    helper::ReadAccessor<Data<helper::vector<unsigned int> >> vPolynomialDegree = d_polynomialDegree;
+
+    msg_info() << "Polynomial derivative data: ";
+    double highOrderStrain = 1.0;
+    double result = 0.0;
+    for (size_t degreeIndex = 0; degreeIndex < vPolynomialDegree[springIndex]; degreeIndex++) {
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:582]: Same comment as previous ones.

File: SofaKernel/modules/SofaDeformable/SofaDeformable_test/PolynomialRestShapeSpringsForceField_test.cpp [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+using sofa::defaulttype::Vec3;
+
+class PolynomialRestShapeSpringsForceField_test : public BaseSimulationTest
+{
+public:
+    /// Run one stepsof simulation then check results
+    bool testSpringForce()
+    {
+        double dt = 0.01;
+        sofa::simpleapi::importPlugin("SofaComponentAll") ;
```
Comment (damienmarchal) [inline SofaKernel/modules/SofaDeformable/SofaDeformable_test/PolynomialRestShapeSpringsForceField_test.cpp:38]: The tests should include corner cases to insure the component is behaving well  (so working in a meaninfull manner and not crashing) when users sets bad inputs in the data field.

**Review by damienmarchal — CHANGES_REQUESTED** [review]
The feature is interesting and I recommand merging...but before doing that it is mandatory to make changes to stick to sofa code standard.

Recommended changes:
- the datafield with similar roles as the ones in SpringsForceField and RestShapeSrpingsForceField should have the same names so users can swtich one component to the other easily. 
- add tests for corner cases and bad cases to enforce component robustness when users are putting wrong values in the data fields. 
- use c++11/c++17 features so the code is more readable (eg: for each loop)
- handle correctly the message for the user with msg_* and the message for the developpers with dmsg_*
- use the if(notMuted) to skips loops that only do msg_info() 
- use a single msg_* instead of a series... much faster and more clear for users in GUI.  
- use getRead/WriteAccessor instead of the automatic conversion version.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl (hunk 16 above)
Comment (hugtalbot) [reply SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:420]: would you have an example to point me out @damienmarchal pliz?

Comment (damienmarchal) [reply SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:420]: It is called  ScopedAdvancedTimer, a search  in the codebase will give you examples.

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl (hunk 18 above)
Comment (hugtalbot) [reply SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:557]: I don't see much usages of this ```getReadAccessor``` in the core of SOFA. Is this normal @damienmarchal ?

File: SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl (hunk 11 above)
Comment (hugtalbot) [reply SofaKernel/modules/SofaDeformable/PolynomialSpringsForceField.inl:381]: there is no drawCylinders function in drawTool and not sure it makes sense to create one..
@fredroy @damienmarchal
